File: persistence/src/main/java/io/sparkled/persistence/sequence/impl/query/GetStageBySequenceIdQuery.java
package io.sparkled.persistence.sequence.impl.query;

import io.sparkled.model.entity.Stage;
import io.sparkled.persistence.PersistenceQuery;
import io.sparkled.persistence.QueryFactory;

import java.util.Optional;

public class GetStageBySequenceIdQuery implements PersistenceQuery<Optional<Stage>> {

    private final int sequenceId;

    public GetStageBySequenceIdQuery(int sequenceId) {
        this.sequenceId = sequenceId;
    }

    @Override
    public Optional<Stage> perform(QueryFactory queryFactory) {
        Stage stage = queryFactory
                .selectFrom(qStage)
                .innerJoin(qSequence).on(qSequence.stageId.eq(qStage.id))
                .where(qSequence.id.eq(sequenceId))
                .fetchFirst();

        return Optional.ofNullable(stage);
    }
}

File: persistence/src/main/kotlin/io/sparkled/persistence/sequence/impl/query/GetStageBySequenceIdQuery.kt
package io.sparkled.persistence.sequence.impl.query

import io.sparkled.model.entity.QSequence.Companion.sequence
import io.sparkled.model.entity.QStage.Companion.stage
import io.sparkled.model.entity.Stage
import io.sparkled.persistence.PersistenceQuery
import io.sparkled.persistence.QueryFactory

class GetStageBySequenceIdQuery(private val sequenceId: Int) : PersistenceQuery<Stage?> {

    override fun perform(queryFactory: QueryFactory): Stage? {
        return queryFactory
            .selectFrom(stage)
            .innerJoin(sequence).on(sequence.stageId.eq(stage.id))
            .where(sequence.id.eq(sequenceId))
            .fetchFirst()
    }
}
